adrs/0002-kubernetes-native-api.md

```diff
@@ -0,0 +1,589 @@
+   are immutable).
+1. The user can opt-in to the Kubernetes native APIs and is not forced to use them or migrate their existing pipeline
+   definitions to the Kubernetes native API feature.
+1. Provide a sample migration script for users wanting to migrate their existing pipelines and pipeline versions to the
+   Kubernetes native API approach.
+1. The Kubeflow Pipelines Python SDK is enhanced to allow the output format to be a Kubernetes native API manifest.
+
+### Non-Goals
+
+1. Kubeflow pipeline runs will not be included in the Kubernetes native API at this time. It is a desire to eventually
+   do this, but it has scale concerns (e.g. 80k+ runs by some users) that require a different approach with tradeoffs.
+1. This will not support the v1 API of Kubeflow pipelines. To continue using the v1 API, the user must continue using
+   the default database solution.
+
+## Proposal
+
+The proposal is to allow the user to opt-in to a backwards compatible Kubernetes native API, achieved through custom
+resource definitions (CRDs), to define pipelines and pipeline versions. The proposed approach is to use the Kubernetes
+API as the pipeline storage interface in the REST API to make things seamless and backwards compatible while not
+introducing data duplication or synchronization issues.
+
+The same validation and restrictions would apply to both the Kubernetes API and the REST API such as pipeline versions
+being immutable, needing to reference an existing pipeline, and have a valid pipeline spec. This will be achieved
+through the API server having endpoints for Kubernetes validating and mutating webhooks that leverage the same Go code
+used by the API server. This will ensure that there is no version drift of the validation logic between the REST API and
+Kubernetes API.
+
+In multiuser mode, the existing RBAC for the REST API performs a subject access review to see if the user has access to
+the `pipelines` resource in the `pipelines.kubeflow.org` API group in the Kubernetes namespace. This will be the same
+requirement on the Kubernetes native API except that the user will additionally need to have access to the
+`pipelineversions` resource. When the Kubernetes native API is enabled on the Kubeflow Pipelines API server, the user
+will also need access to the `pipelineversions` resource when leveraging the REST API. When the user has not opted in to
+the Kubernetes native API, there is no change.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a Kubeflow Pipelines user, I can leverage GitOps tool such as Argo CD to define a pipeline and pipeline version.
+
+#### Story 2
+
+As a Kubeflow Pipelines user, I can use either the REST API or Kubernetes API to see available pipelines.
+
+### Notes/Constraints/Caveats (Optional)
+
+#### Caveats
+
+Migrations from storing the pipelines and pipeline versions in the database to the Kubernetes API will lead to the same
```

Comment:
Would it be worth adding an annotation on the pipeline and pipeline version to put the old UID so that the UI will still map to the proper pipeline version? My concern is that the UID wouldn't really be a unique identifier anymore.